tests/test_interactions.py

```diff
@@ -0,0 +1,153 @@
+            :param y: initial position of the node at the start of the action chain
+            :param delta_x: how much we want to drag the node
+            :param delta_y: how much we want to drag the node
+            :param dir_name: The directory in which we store our screenshots
+            :return: the difference between the position after drag and the starting position
+            """
+            actions.reset_actions()
+            actions.move_to_element_with_offset(
+                self.driver.find_element_by_tag_name('body'), x, y
+            )
+            actions.drag_and_drop_by_offset(source=None, xoffset=delta_x, yoffset=delta_y)
+            actions.click()
+            actions.perform()
+            time.sleep(1)
+
+            elem_json = json.loads(elem.text)
+            new_pos = elem_json.get('renderedPosition')
+            clicked_label = elem_json.get('data', {}).get('label')
+
+            diff_x = round(new_pos['x'] - x)
+            diff_y = round(new_pos['y'] - y)
+
+            save_screenshot(
+                dir_name,
+                'Dragged{}By{}x{}y'.format(clicked_label.replace(' ', ''), diff_x, diff_y)
+            )
+
+            return diff_x, diff_y
+
+        def perform_clicking(x, y, elem, dir_name='interactions'):
+            """
+            :param x: The position on the screen where we want to click
+            :param y: The position on the screen where we want to click
+            :param elem: The element object from where we retrieve the JSON
+            :param dir_name: The directory in which we store our screenshots
+            :return: The label of element most recently clicked, if any
+            """
+            actions.reset_actions()
+            actions.move_to_element_with_offset(
+                self.driver.find_element_by_tag_name('body'), x, y
+            )
+            actions.click()
+            actions.perform()
+
+            time.sleep(1)
+            clicked_label = json.loads(elem.text).get('data', {}).get('label')
+
+            save_screenshot(dir_name, 'Clicked' + clicked_label.replace(' ', ''))
+
+            return clicked_label
+
+        def perform_mouseover(x, y, elem, dir_name='interactions'):
+            actions.reset_actions()
+            actions.move_to_element_with_offset(
+                self.driver.find_element_by_tag_name('body'), x - 100, y
+            )
+            actions.move_by_offset(100, 0)
+            actions.perform()
+            time.sleep(1)
+
+            mouseover_label = json.loads(elem.text).get('label')
+
+            save_screenshot(dir_name, 'Mouseover' + mouseover_label.replace(' ', ''))
+
+            return mouseover_label
+
+        drag_error = "Unable to drag Cytoscape nodes properly"
+        click_error = "Unable to click Cytoscape nodes properly"
+        mouseover_error = "Unable to mouseover Cytoscape nodes properly"
+
+        init_pos = {
+            'Node 1': (80.94611044209678, 333.54879281525285),
```

Comment:
how did you determine these numbers? Could you add a comment?

Reply:
I had to manually find those. It's impossible to get the initial **rendered** positions of the nodes, since we would need to obtain the node object (in dictionary format), which can be done either by (a) using the *tapNode* callback, which paradoxally requires you to click the node, or (b) by making a complex calculation relative to the size of the screen w.r.t largest coordinate in the list of elements. But (b) is unreliable, since I do not know how much padding around the graph is required, so it will likely be off.

I'll add a note about this.

Reply:
Cool - it'd be good to know _how_ you manually found them just in case someone else needs to add some tests as well. The numbers that you have are extremely precise, so it seems like you had some process in place.

Reply:
After setting the size of the window to 1280x1000, I manually clicked on the nodes and used the JSON output from the GUI to collect the number. You are totally right on this: I will make a note so that when the test will be update this can be easily reproduced.